### PR TITLE
i#1557: cmake disallows LOCATION target property, CMP0026

### PR DESCRIPTION
**Pull request by hgreving2304**
Incremental re-factoring, eliminating uses of LOCATION, but not all of them. Reading the target property is replaced by $<TARGET_FILE:target> and file(GENERATE ..) of the former.

Issue: #1557

### CODE REVIEW

File: clients/drcachesim/CMakeLists.txt [hunk 1]
```diff
@@ -407,16 +404,24 @@ endif (X64)
 set(BUILD_CLIENTS_BIN clients/${INSTALL_BIN})
 set(BUILD_CLIENTS_LIB clients/${INSTALL_LIB})
 
-write_config_file(${CONFIG_INSTALL} ${INSTALL_CLIENTS_BIN} ${INSTALL_CLIENTS_LIB})
-write_config_file(${CONFIG_BUILD} ${BUILD_CLIENTS_BIN} ${BUILD_CLIENTS_LIB})
+if (DEBUG)
+  set(debugopt "TOOL_OP=-dr_debug")
```
Comment (derekbruening) [inline clients/drcachesim/CMakeLists.txt]: It seems nicer to keep all of the config file syntax inside write_config_file()

File: clients/drcachesim/CMakeLists.txt [hunk 2]
```diff
@@ -583,7 +586,7 @@ add_custom_command(
   COMMAND ${CMAKE_COMMAND}
   ARGS -D src=${srcdoc}
        -D dst=${gendoc}
-       -D prog=${ops_path}
+       -D prog=$<TARGET_FILE:drcachesim_ops>${location_suffix}
```
Comment (derekbruening) [inline clients/drcachesim/CMakeLists.txt]: This doesn't look right: we don't want "_Debug" on the end of the path.  How do we specify the config type to this $<TARGET_FILE:xxx> syntax?

Same thing below as well.

File: core/CMake_globalize_pic_thunks.cmake [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+# DAMAGE.
+
+# Caller must set:
+# + "lib_fileloc" to point to
```
Comment (derekbruening) [inline core/CMake_globalize_pic_thunks.cmake]: Point to what?

File: core/CMake_readelf.cmake [hunk 4]
```diff
@@ -40,9 +40,16 @@
 # PR 212290: ensure no text relocations (they violate selinux execmod policies)
 # looking for dynamic section tag:
 #   0x00000016 (TEXTREL)                    0x0
+
+include(../drpreloadloc.cmake)
```
Comment (derekbruening) [inline core/CMake_readelf.cmake]: Hmm, I thought this would take in the file to include, so we don't have to hardcode anything here.

File: core/CMakeLists.txt [hunk 5]
```diff
@@ -624,7 +626,7 @@ if (UNIX)
         POST_BUILD
         COMMAND ${CMAKE_COMMAND}
         # to work around i#84 be sure to put a space after -D for 1st arg at least
-        ARGS -D lib=${lib_to_check}
+        ARGS -D lib_fileloc=${locvar_name}
```
Comment (derekbruening) [inline core/CMakeLists.txt]: See below: I thought we'd pass in the ${locavar_name}.cmake path so that CMake_readelf.cmake can include it

File: core/CMakeLists.txt [hunk 6]
```diff
@@ -682,11 +684,13 @@ else (UNIX)
   else (X64)
     set(LD_FLAGS "${LD_FLAGS} /machine:I386")
   endif (X64)
+
```
Comment (derekbruening) [inline core/CMakeLists.txt]: Unintended change?

File: core/CMakeLists.txt [hunk 7]
```diff
@@ -767,27 +772,12 @@ if (NOT APPLE)
...
-        --globalize-symbol=__x86.get_pc_thunk.cx
-        --globalize-symbol=__x86.get_pc_thunk.dx
-        --globalize-symbol=__x86.get_pc_thunk.di
-        --globalize-symbol=__x86.get_pc_thunk.si
-        --globalize-symbol=__x86.get_pc_thunk.bp
-        "${dynamorio_dot_o}" &&
-        )
-    endif ()
+
+    set(locvar_name dynamoriosloc)
```
Comment (derekbruening) [inline core/CMakeLists.txt]: Having a hard time parsing this variable name: maybe add underscores or rename

File: core/CMake_globalize_pic_thunks.cmake [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+execute_process(COMMAND
+  ${CMAKE_C_COMPILER} ${partial_link_flags} -fPIC
+  -nostartfiles -nodefaultlibs -Wl,-r ${disable_pie_flag}
+  -Wl,--whole-archive "${dynamorio_dot_a}" -Wl,--no-whole-archive
+  -o ${dynamorio_dot_o}
+  RESULT_VARIABLE cmd_result
+  ERROR_VARIABLE cmd_error
+  )
+
+if (cmd_error)
```
Comment (derekbruening) [inline core/CMake_globalize_pic_thunks.cmake]: Not checking cmd_result: are we sure it never fails w/o printing sthg?  Ditto below: seems safest to always check the result code.

File: make/utils_exposed.cmake [hunk 9]
```diff
@@ -80,14 +80,20 @@ function (DynamoRIO_add_rel_rpaths target)
...
     foreach (lib ${ARGN})
       # Compute the relative path between the directory of the target and the
       # library it is linked against.
-      get_target_property(tgt_path ${target} LOCATION)
-      get_target_property(lib_path ${lib} LOCATION)
-      _DR_dirname(tgt_path "${tgt_path}")
-      _DR_dirname(lib_path "${lib_path}")
-      file(RELATIVE_PATH relpath "${tgt_path}" "${lib_path}")
+
+      # TODO is this here somehow needed for win at all? / shouldn't, no rpath
```
Comment (derekbruening) [inline make/utils_exposed.cmake]: style again

File: make/utils_exposed.cmake [hunk 10]
```diff
@@ -138,6 +144,10 @@ function (_DR_check_if_linker_is_gnu_gold var_out)
 endfunction (_DR_check_if_linker_is_gnu_gold)
 
 function (DynamoRIO_get_target_path_for_execution out target device_base_dir)
+  # TODO: various caller functions depends on abspath of target at configure time.
```
Comment (derekbruening) [inline make/utils_exposed.cmake]: style

grammar: s/depends/depend/

File: make/utils_exposed.cmake [hunk 11]
```diff
@@ -138,6 +144,10 @@ function (_DR_check_if_linker_is_gnu_gold var_out)
...
+  # If we want to eliminate LOCATION, We need to re-factor all the calls for all
```
Comment (derekbruening) [inline make/utils_exposed.cmake]: nit: grammar: s/We/we/

File: suite/tests/CMakeLists.txt [hunk 12]
```diff
@@ -1595,6 +1593,12 @@ if (X86) # FIXME i#1551, i#1569: port asm to ARM and AArch64
     # XXX i#1651: we don't support -native_exec combined with -early_inject so
     # for now we pass -no_early_inject to all of these tests.
     tobuild_appdll(common.nativeexec common/nativeexec.c)
+
+    # TODO: CMP0026/comment: This LOCATION call depends on configuration time used by either
```
Comment (derekbruening) [inline suite/tests/CMakeLists.txt]: style

more below

File: .travis.yml [hunk 13]
```diff
@@ -95,7 +95,7 @@ jobs:
       env: DYNAMORIO_CROSS_AARCHXX_LINUX_ONLY=no DEPLOY=no EXTRA_ARGS=32_only
       # XXX i#2764: Travis OSX resources are over-subscribed and it can take
       # hours to get an OSX machine, so we skip running PR's for now.
-      if: type = push
+      # if: type = push
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply .travis.yml]: Ok

File: api/samples/CMakeLists.txt [hunk 14]
```diff
@@ -162,6 +162,7 @@ function (add_sample_client name source_file_list extension_list)
 
   # Provide a hint for how to use the client
   if (NOT DynamoRIO_INTERNAL OR NOT "${CMAKE_GENERATOR}" MATCHES "Ninja")
+    # TODO, replace w/ generator expression, CMP0026
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply api/samples/CMakeLists.txt]: Ok

File: api/samples/CMakeLists.txt [hunk 15]
```diff
@@ -314,6 +315,7 @@ DR_install(FILES "${public_clist}"
   DESTINATION "${INSTALL_SAMPLES}" RENAME CMakeLists.txt)
 if (WIN32)                                                    # NON-PUBLIC
   # have a copy of dynamorio library for tracedump            # NON-PUBLIC
+  # TODO, replace w/ generator expression, CMP0026
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply api/samples/CMakeLists.txt]: Ok

[the remark this answers is not in the document]
Comment (hgreving2304) [reply api/samples/CMakeLists.txt]: Done

File: clients/drcachesim/CMakeLists.txt (hunk 1 above)
Comment (hgreving2304) [reply clients/drcachesim/CMakeLists.txt]: Done

File: core/CMake_globalize_pic_thunks.cmake (hunk 8 above)
Comment (hgreving2304) [reply core/CMake_globalize_pic_thunks.cmake]: Done

File: suite/tests/CMakeLists.txt (hunk 12 above)
Comment (hgreving2304) [reply suite/tests/CMakeLists.txt]: Done

**Comment by hgreving2304** [comment]
Are we good here?

**Comment by derekbruening** [comment]
> Are we good here?

It is hard to tell when it's ready for another look -- maybe best to put a "PTAL" comment here once review changes are all responded to.  Taking a look now.

File: api/samples/CMakeLists.txt [hunk 16]
```diff
@@ -312,13 +313,14 @@ DR_install(TARGETS ${tgts} DESTINATION ${INSTALL_SAMPLES_BIN})
...
   DESTINATION "${INSTALL_SAMPLES}" RENAME CMakeLists.txt)
-if (WIN32)                                                    # NON-PUBLIC
-  # have a copy of dynamorio library for tracedump            # NON-PUBLIC
-  get_target_property(DR_TARGET_LOCATION dynamorio            # NON-PUBLIC
-    LOCATION${location_suffix})                               # NON-PUBLIC
-  DR_install(FILES "${DR_TARGET_LOCATION}"                    # NON-PUBLIC
-    DESTINATION "${INSTALL_SAMPLES_BIN}")                     # NON-PUBLIC
-endif (WIN32)                                                 # NON-PUBLIC
+if (WIN32)                                                         # NON-PUBLIC
+  # have a copy of dynamorio library for tracedump                 # NON-PUBLIC
```
Comment (derekbruening) [inline api/samples/CMakeLists.txt]: style: Please add cap + end punct to clearly separate from next line

File: clients/drcachesim/CMakeLists.txt (hunk 2 above)
Comment (derekbruening) [reply clients/drcachesim/CMakeLists.txt]: I don't see a reply here: what does this turn into for VS?

File: core/CMakeLists.txt [hunk 17]
```diff
@@ -67,6 +67,12 @@ if (AARCH64)
   include_directories(BEFORE ${PROJECT_BINARY_DIR})
 endif ()
 
+set(locvar_rel "")
+if ("${CMAKE_VERSION}" VERSION_EQUAL "3.10" OR
+    "${CMAKE_VERSION}" VERSION_GREATER "3.10")
+  set(locvar_rel "../")
```
Comment (derekbruening) [inline core/CMakeLists.txt]: I don't understand why the version changes the path: please add a comment explaining

File: core/CMakeLists.txt (hunk 6 above)
Comment (derekbruening) [reply core/CMakeLists.txt]: I don't see a reply: generally these would be reverted

File: core/CMakeLists.txt (hunk 7 above)
Comment (derekbruening) [reply core/CMakeLists.txt]: No reply here either: same reaction to "dynamoriosloc"

File: core/CMake_globalize_pic_thunks.cmake (hunk 3 above)
Comment (derekbruening) [reply core/CMake_globalize_pic_thunks.cmake]: No reply or action here

File: core/CMake_globalize_pic_thunks.cmake [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+# Caller must set:
+# + "lib_fileloc" to point to
+# + CMAKE_C_COMPILER to point to CC compiler
+# + partial_link_flags to linker flags incl. -m32/-m64
+# + disable_pie_flag to -no-pie or ""
+# + CMAKE_OBJCOPY to point to objcopy
+# + CMAKE_AR to point to ar
+# + CMAKE_RANLIB to point to ranlib
+
+include(../../dynamoriosloc.cmake)
```
Comment (derekbruening) [inline core/CMake_globalize_pic_thunks.cmake]: I thought the idea was to pass in the file name to include?

File: clients/drcachesim/CMakeLists.txt (hunk 2 above)
Comment (hgreving2304) [reply clients/drcachesim/CMakeLists.txt]: These and other comments below were hidden, didn't see, will address.

Comment (derekbruening) [reply clients/drcachesim/CMakeLists.txt]: Do we understand how they were hidden in your view?  Did you commit changes to the same lines of code where the comments were before looking at all of the comments?  I think that is when Github marks as "outdated" or sthg.  Is that what you mean by "hidden", still there but collapsed?

File: core/CMake_globalize_pic_thunks.cmake (hunk 3 above)
Comment (hgreving2304) [reply core/CMake_globalize_pic_thunks.cmake]: Fixed.

File: clients/drcachesim/CMakeLists.txt (hunk 2 above)
Comment (hgreving2304) [reply clients/drcachesim/CMakeLists.txt]: No problem with GitHub, they were just collapsed with a "hidden" link and I thought they were resolved (or otherwise not be collapsed).

File: core/CMake_readelf.cmake (hunk 4 above)
Comment (hgreving2304) [reply core/CMake_readelf.cmake]: Yes we can use ${lib_fileloc}.cmake.

File: core/CMakeLists.txt (hunk 6 above)
Comment (hgreving2304) [reply core/CMakeLists.txt]: Done

File: core/CMakeLists.txt (hunk 5 above)
Comment (hgreving2304) [reply core/CMakeLists.txt]: Done, see above.

File: core/CMakeLists.txt (hunk 7 above)
Comment (hgreving2304) [reply core/CMakeLists.txt]: Renamed to dynamoriostaticloc

File: .travis.yml [hunk 19]
```diff
@@ -95,7 +95,8 @@ jobs:
       env: DYNAMORIO_CROSS_AARCHXX_LINUX_ONLY=no DEPLOY=no EXTRA_ARGS=32_only
       # XXX i#2764: Travis OSX resources are over-subscribed and it can take
       # hours to get an OSX machine, so we skip running PR's for now.
-      if: type = push
+      # *** TEMPORARY, will get removed before commit ***
+      # if: type = push
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply .travis.yml]: I had left this in intentionally so Mac is getting tested. Will do before final commit?

File: api/samples/CMakeLists.txt (hunk 16 above)
Comment (hgreving2304) [reply api/samples/CMakeLists.txt]: Done

File: core/CMakeLists.txt (hunk 17 above)
Comment (hgreving2304) [reply core/CMakeLists.txt]: CMP0070 is the policy. They changed the behavior to include the relative path. One could disable CMP0070, but I suppose we don't want to add more "OLD" policies, rather we're adding the "NEW" version.

File: core/CMake_globalize_pic_thunks.cmake (hunk 18 above)
Comment (hgreving2304) [reply core/CMake_globalize_pic_thunks.cmake]: Fixed, I removed the fileloc_var for the _globalize script. It's always dynamoriostatic.

File: .travis.yml (hunk 19 above)
[the remark this answers is not in the document]
Comment (derekbruening) [reply .travis.yml]: OK

File: core/CMakeLists.txt (hunk 7 above)
Comment (derekbruening) [reply core/CMakeLists.txt]: Please use underscores: that is very hard to parse

File: clients/drcachesim/CMakeLists.txt (hunk 2 above)
Comment (hgreving2304) [reply clients/drcachesim/CMakeLists.txt]: The old ops_path also contained ${location_suffix}
Is there a AppVeyor test for building documentation under Windows?

Comment (derekbruening) [reply clients/drcachesim/CMakeLists.txt]: The old ops_path did this:

```
get_target_property(ops_path drcachesim_ops LOCATION${location_suffix})
```

That turns into the LOCATION_DEBUG or LOCATION_RELWITHDEBINFO property.  I.e., ${location_suffix} is on the property name, not on the path.

Comment (hgreving2304) [reply clients/drcachesim/CMakeLists.txt]: I didn't even know LOCATION_DEBUG etc. exists. I'm able to build Win docs now, checking this out WIP.

File: .travis.yml (hunk 19 above)
[the remark this answers is not in the document]
Comment (hgreving2304) [reply .travis.yml]: Removed

File: clients/drcachesim/CMakeLists.txt (hunk 2 above)
Comment (hgreving2304) [reply clients/drcachesim/CMakeLists.txt]: I verified on Windows that TARGET_FILE seems to reflect the correct name.

**Comment by hgreving2304** [comment]
(done). PTAL

File: clients/drcachesim/CMakeLists.txt (hunk 2 above)
Comment (derekbruening) [reply clients/drcachesim/CMakeLists.txt]: So `$<TARGET_FILE:drcachesim_ops>_DEBUG` is the correct syntax?  It seems weird: I'd expect `$<TARGET_FILE_DEBUG:drcachesim_ops>`.

Comment (hgreving2304) [reply clients/drcachesim/CMakeLists.txt]: No it's like you said, the branch here was updated to remove the suffix.

Comment (derekbruening) [reply clients/drcachesim/CMakeLists.txt]: Who adds the configuration-specific subdir for multi-configuration generators like VS or XCode, then, where a single CMake config can build either Debug or Relwithdebinfo and uses a subdir in the build dir?  Is that built in to generator expressions somehow?  DR collapses them to one, but still had to use the LOCATION_xxxx b/c plain LOCATION was empty IIRC for VS generators -- are we sure these TARGET_FILE generator expressions do the right thing?  Can you point at any docs on this?  I couldn't find good info on a quick search.

Comment (hgreving2304) [reply clients/drcachesim/CMakeLists.txt]: Yes so I can't answer this yet as well. What I did was to check whether we're building the same way (i.e. prog has the same value) in Windows. I am assuming the generator expression provides the final correct path dependent on what you're building. I checked RelWithDebInfo and Debug and both times the expression had the right path. I'll see if I can find some docs

**Comment by hgreving2304** [comment]
PTAL

File: clients/drcachesim/CMakeLists.txt (hunk 2 above)
Comment (hgreving2304) [reply clients/drcachesim/CMakeLists.txt]: Mailing list replied that TARGET_FILE generator expression does give the per-config location for multi-config generators.

File: .travis.yml [hunk 20]
```diff
@@ -95,6 +95,7 @@ jobs:
       env: DYNAMORIO_CROSS_AARCHXX_LINUX_ONLY=no DEPLOY=no EXTRA_ARGS=32_only
       # XXX i#2764: Travis OSX resources are over-subscribed and it can take
       # hours to get an OSX machine, so we skip running PR's for now.
+      # *** TEMPORARY, will get removed before commit ***
```
Comment (derekbruening) [inline .travis.yml]: You forgot to remove this?

File: core/CMake_globalize_pic_thunks.cmake [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+
+# Caller must set:
+# + CMAKE_C_COMPILER to point to CC compiler
+# + partial_link_flags to linker flags incl. -m32/-m64
+# + disable_pie_flag to -no-pie or ""
+# + CMAKE_OBJCOPY to point to objcopy
+# + CMAKE_AR to point to ar
+# + CMAKE_RANLIB to point to ranlib
+
+include(../../dynamorio_static_loc.cmake)
```
Comment (derekbruening) [inline core/CMake_globalize_pic_thunks.cmake]: This is still hardcoded?  I thought we agreed to parameterize this?  You're passing in the parameter, but it's not listed in the inputs and it's not used here.

File: core/CMake_readelf.cmake [hunk 22]
```diff
@@ -40,9 +40,12 @@
 # PR 212290: ensure no text relocations (they violate selinux execmod policies)
 # looking for dynamic section tag:
 #   0x00000016 (TEXTREL)                    0x0
+
+include(../${lib_fileloc}.cmake)
```
Comment (derekbruening) [inline core/CMake_readelf.cmake]: Please list among the inputs "Caller must set" above.

Make it clear this is both the file name *and* the variable name that the file sets.  That seems weird: I would expect the file name to be an absolute path.  Explain precisely what kind of path and what assumptions on where it's located.  Ditto for all the other .cmake files taking in "lib_fileloc".

File: make/policies.cmake [hunk 23]
```diff
@@ -34,20 +34,26 @@ if ("${CMAKE_VERSION}" VERSION_EQUAL "3.3" OR
...
-# XXX i#1418: update to cmake 2.8.12's better handling of interface imports, for Qt
+# XXX i#1418: update to cmake 2.8.12's better handling of interface imports, for Qt/Win
 cmake_policy(SET CMP0020 OLD)
 
 # Recognize literals in if statements
 cmake_policy(SET CMP0012 NEW)
+
+if ("${CMAKE_VERSION}" VERSION_EQUAL "3.10" OR
+    "${CMAKE_VERSION}" VERSION_GREATER "3.10")
+  cmake_policy(SET CMP0070 NEW)
```
Comment (derekbruening) [inline make/policies.cmake]: Worth a comment explaining what this is

File: make/utils_exposed.cmake [hunk 24]
```diff
@@ -80,14 +80,20 @@ function (DynamoRIO_add_rel_rpaths target)
...
     foreach (lib ${ARGN})
       # Compute the relative path between the directory of the target and the
       # library it is linked against.
-      get_target_property(tgt_path ${target} LOCATION)
-      get_target_property(lib_path ${lib} LOCATION)
-      _DR_dirname(tgt_path "${tgt_path}")
-      _DR_dirname(lib_path "${lib_path}")
-      file(RELATIVE_PATH relpath "${tgt_path}" "${lib_path}")
+
+      # XXX: is this here needed for win at all? It shouldn't, no rpath
```
Comment (derekbruening) [inline make/utils_exposed.cmake]: Remove this line (and the blank line): the answer is no and Windows is already excluded by line 80 above.

File: make/utils_exposed.cmake [hunk 25]
```diff
@@ -80,14 +80,20 @@ function (DynamoRIO_add_rel_rpaths target)
...
+      get_target_property(target_type ${target} TYPE)
+      if (target_type STREQUAL "EXECUTABLE")
+        get_target_property(target_loc ${target} RUNTIME_OUTPUT_DIRECTORY)
+      else ()
+        get_target_property(target_loc ${target} LIBRARY_OUTPUT_DIRECTORY)
+      endif()
+      get_target_property(lib_loc ${lib} LIBRARY_OUTPUT_DIRECTORY)
```
Comment (derekbruening) [inline make/utils_exposed.cmake]: This is worth a comment that it assumes targets are in standard locations, that it does *not* look at the actual target's destination, and it's doing it all b/c of CMake no longer supporting getting locations at config time.  Help the reader understand why it's doing it this way.

File: make/utils_exposed.cmake [hunk 26]
```diff
@@ -138,6 +144,10 @@ function (_DR_check_if_linker_is_gnu_gold var_out)
 endfunction (_DR_check_if_linker_is_gnu_gold)
 
 function (DynamoRIO_get_target_path_for_execution out target device_base_dir)
+  # XXX: various caller functions depend on abspath of target at configure time.
```
Comment (derekbruening) [inline make/utils_exposed.cmake]: i#1557

File: tools/CMakeLists.txt [hunk 27]
```diff
@@ -182,11 +182,17 @@ foreach (deploytgt drconfig drrun drinject)
     target_link_libraries(${deploytgt} drinjectlib)
   endif ()
   if (LINUX) # No checks for MacOS
-    get_target_property(check_path ${deploytgt} LOCATION)
-    add_custom_command(TARGET ${deploytgt}
-      POST_BUILD
+    set(locvar_rel "")
+    if ("${CMAKE_VERSION}" VERSION_EQUAL "3.10" OR
+        "${CMAKE_VERSION}" VERSION_GREATER "3.10")
+      set(locvar_rel "../")
```
Comment (derekbruening) [inline tools/CMakeLists.txt]: This is being used multiple times: I think this should be in a shared function so we have the comment explaining what it is in one place.  Maybe the function could include the file(GENERATE if that's easy to do.

File: core/CMakeLists.txt [hunk 28]
```diff
@@ -796,19 +795,21 @@ if (NOT APPLE)
...
-      ${CMAKE_OBJCOPY} --localize-hidden "${dynamorio_dot_o}" &&
-      ${globalize_pic_thunks}
-      rm -f "${dynamorio_dot_a}" &&
-      ${CMAKE_AR} cr "${dynamorio_dot_a}" "${dynamorio_dot_o}" &&
-      ${CMAKE_RANLIB} "${dynamorio_dot_a}"
+
+    add_custom_command(TARGET dynamorio_static
+      POST_BUILD
+      COMMAND ${CMAKE_COMMAND}
+      ARGS -D lib_fileloc=${locvar_name}
```
Comment (derekbruening) [inline core/CMakeLists.txt]: Below this sets WORKING_DIRECTORY to the lib output dir, yet the locvar_name.cmake is generated wrt the current build dir: it seems like it's going to fail if we use relative dirs.  It seems very fragile.  Wouldn't absolute dirs be more robust and have fewer problems?

File: tools/CMakeLists.txt [hunk 29]
```diff
@@ -182,11 +182,17 @@ foreach (deploytgt drconfig drrun drinject)
...
+    endif()
+    set(locvar_name ${deploytgt}loc)
+    file(GENERATE OUTPUT "${locvar_rel}${locvar_name}.cmake" CONTENT "set(${locvar_name} \
+      \"$<TARGET_FILE:${deploytgt}>\")\n")
+    add_custom_command(TARGET ${deploytgt} POST_BUILD
       COMMAND ${CMAKE_COMMAND}
-      ARGS -D lib=${check_path}
+      ARGS -D lib_fileloc=${locvar_name}
```
Comment (derekbruening) [inline tools/CMakeLists.txt]: The relative dirs used make me nervous...I guess I'm echoing the comment up above.  So you're relying on the file generation and the script to have the same working dir, the current binary dir, so if we call CMake_readelf.cmake from tools/ and from core/ the file generation and the script are both in that subdir I suppose.  But any WORKING_DIRECTORY param seems like it will break it.  Absolute seems more robust.  Just repeating my above comment I guess.

File: .travis.yml (hunk 20 above)
Comment (hgreving2304) [reply .travis.yml]: Will remove before commit.

Comment (derekbruening) [reply .travis.yml]: Friendly suggestion: please look through the patch yourself beforehand :)

File: core/CMake_globalize_pic_thunks.cmake (hunk 21 above)
Comment (hgreving2304) [reply core/CMake_globalize_pic_thunks.cmake]: I had left it got dynamorio_static, because it's always the same, I've changed it to the variable, done/

File: core/CMake_readelf.cmake (hunk 22 above)
Comment (hgreving2304) [reply core/CMake_readelf.cmake]: Done.

File: make/utils_exposed.cmake (hunk 25 above)
Comment (hgreving2304) [reply make/utils_exposed.cmake]: Done

File: core/CMake_globalize_pic_thunks.cmake (hunk 21 above)
Comment (derekbruening) [reply core/CMake_globalize_pic_thunks.cmake]: > I had left it got dynamorio_static, because it's always the same

But if your intent was to leave it hardcoded why was the caller still setting the variable?  It ends up looking like a mistake to the reader.  I would vote for the variable in any case: local vars are better than global vars or hardcoded constants, in general.

Comment (hgreving2304) [reply core/CMake_globalize_pic_thunks.cmake]: Changed to variable already (not uploaded yet)

File: tools/CMakeLists.txt (hunk 27 above)
Comment (hgreving2304) [reply tools/CMakeLists.txt]: The relative path depends on where the caller is located though. So putting it in a function (though we only need '../') makes it less flexible, unlessI'd add some weird parameter like "levels_needed".

It's only need in tools/CMakeLists.txt and core/CMakeLists.txt.

you ok with leaving it as is for now?

File: core/CMakeLists.txt (hunk 28 above)
Comment (hgreving2304) [reply core/CMakeLists.txt]: Do you mean absolute dirs with file(GENERATE) by setting CMP0070 to OLD? I did not want to introduce another 'OLD' behavior policy as the intention of this patch was towards getting rid of them.

File: tools/CMakeLists.txt (hunk 29 above)
Comment (hgreving2304) [reply tools/CMakeLists.txt]: Same question above. Do you mean CMP0070 related?

File: tools/CMakeLists.txt (hunk 27 above)
Comment (derekbruening) [reply tools/CMakeLists.txt]: If we use absolute paths we can eliminate all of this cmake-version-specific code.  We could then have a function that invokes file(GENERATE if that seems useful to avoid duplication, or not if we think the code is small enough once this "../" adjustment disappears.

File: core/CMakeLists.txt (hunk 28 above)
Comment (derekbruening) [reply core/CMakeLists.txt]: Looking at CMP0070 (https://cmake.org/cmake/help/v3.12/policy/CMP0070.html) it only applies when we use a relative path.  So I don't see how CMP0070 comes into play at all if we use absolute paths?  I'm suggesting using ${CMAKE_CURRENT_BINARY_DIR}/myfile or something like that.  Then we remove the fragileness related to where the script's working directory is and what today is an implicit hidden parameter, the current directory.  We also remove any need to care about CMP0070 or adjust any path depending on the cmake version.

Comment (hgreving2304) [reply core/CMakeLists.txt]: Ok, great. When I did this I didn't realize that file(GENERATE) can take absolute paths. I changed it to abs path and removed CMP0070.

File: tools/CMakeLists.txt (hunk 27 above)
Comment (hgreving2304) [reply tools/CMakeLists.txt]: Done

File: core/CMake_globalize_pic_thunks.cmake [hunk 30]
```diff
@@ -30,16 +30,19 @@
 # DAMAGE.
 
 # Caller must set:
+# + "lib_fileloc" to <name> that represents a file <name>.cmake the contains a
```
Comment (derekbruening) [inline core/CMake_globalize_pic_thunks.cmake]: nit: none of the other vars below are in quotes

File: core/CMake_globalize_pic_thunks.cmake [hunk 31]
```diff
@@ -30,16 +30,19 @@
...
+# set(<name> <path>) where path is a library location. The .cmake file must reside
+# in the root build directory.
```
Comment (derekbruening) [inline core/CMake_globalize_pic_thunks.cmake]: nit: indent these 2 lines 2 spaces so the list bullets are clear

Ditto below

File: core/CMake_globalize_pic_thunks.cmake [hunk 32]
```diff
@@ -30,16 +30,19 @@
...
 # + CMAKE_C_COMPILER to point to CC compiler
 # + partial_link_flags to linker flags incl. -m32/-m64
 # + disable_pie_flag to -no-pie or ""
 # + CMAKE_OBJCOPY to point to objcopy
 # + CMAKE_AR to point to ar
 # + CMAKE_RANLIB to point to ranlib
 
-include(../../dynamorio_static_loc.cmake)
+include(../../${lib_fileloc}.cmake)
```
Comment (derekbruening) [inline core/CMake_globalize_pic_thunks.cmake]: But this is still making an assumption on its working directory: how do you know it's always going to be run 2 levels deep?  The point of the absolute path is to avoid these kinds of assumptions.  Just pass in the absolute path and include it directly here.  If you want to keep the filename==varname and not have a separate var for the varname, get the basename of the abs path here as the varname.

Ditto below.

File: make/utils_exposed.cmake [hunk 33]
```diff
@@ -85,13 +85,15 @@ function (DynamoRIO_add_rel_rpaths target)
...
       # library it is linked against.
 
-      # XXX: is this here needed for win at all? It shouldn't, no rpath
       get_target_property(target_type ${target} TYPE)
       if (target_type STREQUAL "EXECUTABLE")
         get_target_property(target_loc ${target} RUNTIME_OUTPUT_DIRECTORY)
       else ()
         get_target_property(target_loc ${target} LIBRARY_OUTPUT_DIRECTORY)
       endif()
+      # Reading the target pathes are configure time is no longer supported in
```
Comment (derekbruening) [inline make/utils_exposed.cmake]: spelling: "paths"

File: core/CMakeLists.txt [hunk 34]
```diff
@@ -614,7 +606,7 @@ if (UNIX)
       message("${readelf_name} not found: not checking SElinux or execstack")
     else ()
       set(locvar_name dynamorioloc)
-      file(GENERATE OUTPUT "${locvar_rel}${locvar_name}.cmake" CONTENT
+      file(GENERATE OUTPUT "${PROJECT_BINARY_DIR}/${locvar_name}.cmake" CONTENT
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply core/CMakeLists.txt]: I intentionally put it there because that's where it was before and it was tested like this. Changed.

File: core/CMake_globalize_pic_thunks.cmake (hunk 32 above)
Comment (hgreving2304) [reply core/CMake_globalize_pic_thunks.cmake]: Done

File: core/CMake_globalize_pic_thunks.cmake [hunk 35]
```diff
@@ -30,19 +30,20 @@
...
-# + partial_link_flags to linker flags incl. -m32/-m64
-# + disable_pie_flag to -no-pie or ""
-# + CMAKE_OBJCOPY to point to objcopy
-# + CMAKE_AR to point to ar
-# + CMAKE_RANLIB to point to ranlib
-
-include(../../${lib_fileloc}.cmake)
-
-set(dynamorio_dot_a "${${lib_fileloc}}")
+# + "lib_fileloc" to /absolute/path/<name> that represents a file <name>.cmake the
```
Comment (derekbruening) [inline core/CMake_globalize_pic_thunks.cmake]: error: s/the/that/

Please read through before sending for review

File: core/CMake_globalize_pic_thunks.cmake [hunk 36]
```diff
@@ -30,19 +30,20 @@
...
+#   contains a set(<name> <path>) where path is a library location.
```
Comment (derekbruening) [inline core/CMake_globalize_pic_thunks.cmake]: suggest: s/a library/the target binary/

File: core/CMake_readelf.cmake [hunk 37]
```diff
@@ -30,10 +30,9 @@
 # DAMAGE.
 
 # Caller must set:
-# + "lib_fileloc" to <name> that represents a file <name>.cmake the contains a
-# set(<name> <path>) where path is a library location. The .cmake file must reside
-# in the root build directory.
-# + READELF_EXECUTABLE so it can be a cache variable
+# + "lib_fileloc" to /absolute/path/<name> that represents a file <name>.cmake the
```
Comment (derekbruening) [inline core/CMake_readelf.cmake]: Same two comments

File: make/utils_exposed.cmake [hunk 38]
```diff
@@ -91,7 +91,7 @@ function (DynamoRIO_add_rel_rpaths target)
       else ()
         get_target_property(target_loc ${target} LIBRARY_OUTPUT_DIRECTORY)
       endif()
-      # Reading the target pathes are configure time is no longer supported in
+      # Reading the target paths are configure time is no longer supported in
```
Comment (derekbruening) [inline make/utils_exposed.cmake]: error: s/are/at/
